In [2]:
from typing import List

import pandas as pd
from pathlib import Path
import os
import sys
from dotenv import load_dotenv, dotenv_values
from pydriller import Repository
import json
from multiprocessing import Pool
from itertools import chain

In [ ]:
load_dotenv()
ENV = dotenv_values(".env")
DATA_DIR = Path(ENV["DATA_DIR"])
DATA_DIR

In [4]:
df = pd.read_csv(DATA_DIR / 'interim' / 'interim_result.1.tsv', delimiter='\t')
df

cveId               ghsaId  \
0     CVE-2019-0214  GHSA-jxgm-9f58-w4xp   
1     CVE-2013-7285  GHSA-f554-x222-wgf7   
2     CVE-2019-0213  GHSA-cqcf-4g4h-rghf   
3    CVE-2019-17513  GHSA-mvqp-q37c-wf9j   
4    CVE-2017-15703  GHSA-xwx6-vmj4-5rv8   
..              ...                  ...   
199   CVE-2020-9296  GHSA-wfj5-2mqr-7jvv   
200  CVE-2020-13692  GHSA-88cc-g835-76rp   
201   CVE-2020-1758  GHSA-c597-f74m-jgc2   
202  CVE-2018-21234  GHSA-jrg3-qq99-35g7   
203  CVE-2020-13973  GHSA-g8jj-899q-8x3j   

                                         packageName vulnIntroduced  \
0                         org.apache.archiva:archiva          2.2.0   
1                   com.thoughtworks.xstream:xstream              0   
2                         org.apache.archiva:archiva              0   
3                            io.ratpack:ratpack-core              0   
4    org.apache.nifi:nifi-framework-cluster-protocol              0   
..                                               ...            ...   
199             com.netflix.conductor:conductor-core              0   
200                        org.postgresql:postgresql              0   
201                     org.keycloak:keycloak-parent              0   
202                                org.jodd:jodd-all              0   
203                    com.mikesamuel:json-sanitizer              0   

    latestVulnVersion vulnFixed                            fixCommitRepo  \
0               2.2.3     2.2.4        https://github.com/apache/archiva   
1               1.4.5     1.4.6      https://github.com/x-stream/xstream   
2               2.2.3     2.2.4        https://github.com/apache/archiva   
3               1.7.4     1.7.5       https://github.com/ratpack/ratpack   
4               1.4.0     1.5.0           https://github.com/apache/nifi   
..                ...       ...                                      ...   
199            2.25.3    2.25.4     https://github.com/Netflix/conductor   
200      42.2.12.jre7   42.2.13         https://github.com/pgjdbc/pgjdbc   
201             9.0.3    10.0.0     https://github.com/keycloak/keycloak   
202             5.0.3     5.0.4            https://github.com/oblac/jodd   
203             1.2.0     1.2.1  https://github.com/OWASP/json-sanitizer   

                                fixCommitHash  
0    796716d44183bd315dd20184a66b39ae533eb747  
1    6344867dce6767af7d0fe34fb393271a6456672d  
2    a36035b49ba7d6514d6c386b51e1ad2512371b3d  
3    efb910d38a96494256f36675ef0e5061097dd77d  
4    9e2c7be7d3c6a380c5f61074d9a5a690b617c3dc  
..                                        ...  
199  375f8d8b708a801647c327e0848b30d34f804edd  
200  14b62aca4764d496813f55a43d050b017e01eb65  
201  97b565469058cacac6d2559b2cff37effcca2cfd  
202  9bffc3913aeb8472c11bb543243004b4b4376f16  
203  f3512bab078af3f5c533dacc2d410add8c0c1124  

[204 rows x 8 columns]

In [5]:
df_dict = df.to_dict(orient='records')

In [11]:
def get_new_records(r) -> List:
    new_records = list()
    repo_path = DATA_DIR / 'interim' / 'repositories'
    repo_path.mkdir(exist_ok=True, parents=True)

    try:
        repo = Repository(r["fixCommitRepo"], single=r["fixCommitHash"], clone_repo_to=repo_path)
    except Exception:
        print(f'error cloning {r["fixCommitRepo"]} ({r["fixCommitHash"]})')
        return []

    try:
        commit = list(repo.traverse_commits())[0]
    except Exception:
        print(f'error getting commit {r["fixCommitRepo"]} ({r["fixCommitHash"]})')
        return []

    # assert commit.hash == r["fixCommitHash"]
    for mf in commit.modified_files:
        try:
            if not mf.filename.endswith('.java') or mf.source_code_before is None:
                continue
            new_row = r.copy()
            new_row['commitHashFromGit'] = commit.hash
            new_row['modifiedFilePathBefore'] = mf.old_path
            new_row['modifiedFilePathAfter'] = mf.new_path
            new_row['modifiedFileSrcBefore'] = mf.source_code_before
            new_row['modifiedFileSrcAfter'] = mf.source_code
            new_row['diffParsedJson'] = json.dumps(mf.diff_parsed)
            new_row['nloc'] = mf.nloc
            new_row['changedMethods'] = list(map(lambda m: m.name, mf.changed_methods))
        
            new_records.append(new_row)

        except Exception:
            print(f'error parsing modified files {r["fixCommitRepo"]} ({r["fixCommitHash"]} ({mf.filename})')
            continue

    return new_records


with Pool(8) as p:
    new_records_list = p.map(get_new_records, df_dict)
    new_df = list(chain(*new_records_list))

pd.DataFrame(data=new_df)

error getting commit https://github.com/apache/poi (3a328aa220f6979f9805f658ae33244d153beaa7)
error getting commit http://svn.apache.org/repos/asf/xerces/java (787352)
error getting commit http://svn.apache.org/repos/asf/activemq (1209700)
error getting commit http://svn.apache.org/repos/asf/httpcomponents/httpclient (1074473)
error getting commit https://github.com/bcgit/bc-java (81b00861cd5711e85fe8dce2a0e119f684120255)


cveId               ghsaId                       packageName  \
0     CVE-2019-0214  GHSA-jxgm-9f58-w4xp        org.apache.archiva:archiva   
1     CVE-2019-0214  GHSA-jxgm-9f58-w4xp        org.apache.archiva:archiva   
2     CVE-2019-0214  GHSA-jxgm-9f58-w4xp        org.apache.archiva:archiva   
3     CVE-2019-0214  GHSA-jxgm-9f58-w4xp        org.apache.archiva:archiva   
4     CVE-2013-7285  GHSA-f554-x222-wgf7  com.thoughtworks.xstream:xstream   
..              ...                  ...                               ...   
645  CVE-2018-21234  GHSA-jrg3-qq99-35g7                 org.jodd:jodd-all   
646  CVE-2018-21234  GHSA-jrg3-qq99-35g7                 org.jodd:jodd-all   
647  CVE-2018-21234  GHSA-jrg3-qq99-35g7                 org.jodd:jodd-all   
648  CVE-2020-13973  GHSA-g8jj-899q-8x3j     com.mikesamuel:json-sanitizer   
649  CVE-2020-13973  GHSA-g8jj-899q-8x3j     com.mikesamuel:json-sanitizer   

    vulnIntroduced latestVulnVersion vulnFixed  \
0            2.2.0             2.2.3     2.2.4   
1            2.2.0             2.2.3     2.2.4   
2            2.2.0             2.2.3     2.2.4   
3            2.2.0             2.2.3     2.2.4   
4                0             1.4.5     1.4.6   
..             ...               ...       ...   
645              0             5.0.3     5.0.4   
646              0             5.0.3     5.0.4   
647              0             5.0.3     5.0.4   
648              0             1.2.0     1.2.1   
649              0             1.2.0     1.2.1   

                               fixCommitRepo  \
0          https://github.com/apache/archiva   
1          https://github.com/apache/archiva   
2          https://github.com/apache/archiva   
3          https://github.com/apache/archiva   
4        https://github.com/x-stream/xstream   
..                                       ...   
645            https://github.com/oblac/jodd   
646            https://github.com/oblac/jodd   
647            https://github.com/oblac/jodd   
648  https://github.com/OWASP/json-sanitizer   
649  https://github.com/OWASP/json-sanitizer   

                                fixCommitHash  \
0    796716d44183bd315dd20184a66b39ae533eb747   
1    796716d44183bd315dd20184a66b39ae533eb747   
2    796716d44183bd315dd20184a66b39ae533eb747   
3    796716d44183bd315dd20184a66b39ae533eb747   
4    6344867dce6767af7d0fe34fb393271a6456672d   
..                                        ...   
645  9bffc3913aeb8472c11bb543243004b4b4376f16   
646  9bffc3913aeb8472c11bb543243004b4b4376f16   
647  9bffc3913aeb8472c11bb543243004b4b4376f16   
648  f3512bab078af3f5c533dacc2d410add8c0c1124   
649  f3512bab078af3f5c533dacc2d410add8c0c1124   

                            commitHashFromGit  \
0    796716d44183bd315dd20184a66b39ae533eb747   
1    796716d44183bd315dd20184a66b39ae533eb747   
2    796716d44183bd315dd20184a66b39ae533eb747   
3    796716d44183bd315dd20184a66b39ae533eb747   
4    6344867dce6767af7d0fe34fb393271a6456672d   
..                                        ...   
645  9bffc3913aeb8472c11bb543243004b4b4376f16   
646  9bffc3913aeb8472c11bb543243004b4b4376f16   
647  9bffc3913aeb8472c11bb543243004b4b4376f16   
648  f3512bab078af3f5c533dacc2d410add8c0c1124   
649  f3512bab078af3f5c533dacc2d410add8c0c1124   

                                modifiedFilePathBefore  \
0    archiva-modules/archiva-base/archiva-repositor...   
1    archiva-modules/archiva-base/archiva-repositor...   
2    archiva-modules/archiva-web/archiva-rest/archi...   
3    archiva-modules/archiva-web/archiva-rest/archi...   
4    xstream/src/java/com/thoughtworks/xstream/XStr...   
..                                                 ...   
645  jodd-json/src/main/java/jodd/json/JsonParserBa...   
646   jodd-json/src/main/java/jodd/json/MapToBean.java   
647  jodd-json/src/test/java/jodd/json/JSONDeserial...   
648   src/main/java/com/google/json/JsonSanitizer.java   
649  src/test/java/com/google/json/JsonSanitizerTes...   

                        

last output:
error getting commit https://github.com/apache/poi (3a328aa220f6979f9805f658ae33244d153beaa7)
error getting commit http://svn.apache.org/repos/asf/xerces/java (787352)
error getting commit http://svn.apache.org/repos/asf/activemq (1209700)
error getting commit http://svn.apache.org/repos/asf/httpcomponents/httpclient (1074473)
error getting commit https://github.com/bcgit/bc-java (81b00861cd5711e85fe8dce2a0e119f684120255)

In [16]:
df2 = pd.DataFrame(data=new_df)
df2 = df2[~df2['modifiedFilePathAfter'].isnull()]
df2 = df2[~df2['modifiedFileSrcAfter'].isnull()]
df2

cveId               ghsaId                       packageName  \
0     CVE-2019-0214  GHSA-jxgm-9f58-w4xp        org.apache.archiva:archiva   
1     CVE-2019-0214  GHSA-jxgm-9f58-w4xp        org.apache.archiva:archiva   
2     CVE-2019-0214  GHSA-jxgm-9f58-w4xp        org.apache.archiva:archiva   
3     CVE-2019-0214  GHSA-jxgm-9f58-w4xp        org.apache.archiva:archiva   
4     CVE-2013-7285  GHSA-f554-x222-wgf7  com.thoughtworks.xstream:xstream   
..              ...                  ...                               ...   
645  CVE-2018-21234  GHSA-jrg3-qq99-35g7                 org.jodd:jodd-all   
646  CVE-2018-21234  GHSA-jrg3-qq99-35g7                 org.jodd:jodd-all   
647  CVE-2018-21234  GHSA-jrg3-qq99-35g7                 org.jodd:jodd-all   
648  CVE-2020-13973  GHSA-g8jj-899q-8x3j     com.mikesamuel:json-sanitizer   
649  CVE-2020-13973  GHSA-g8jj-899q-8x3j     com.mikesamuel:json-sanitizer   

    vulnIntroduced latestVulnVersion vulnFixed  \
0            2.2.0             2.2.3     2.2.4   
1            2.2.0             2.2.3     2.2.4   
2            2.2.0             2.2.3     2.2.4   
3            2.2.0             2.2.3     2.2.4   
4                0             1.4.5     1.4.6   
..             ...               ...       ...   
645              0             5.0.3     5.0.4   
646              0             5.0.3     5.0.4   
647              0             5.0.3     5.0.4   
648              0             1.2.0     1.2.1   
649              0             1.2.0     1.2.1   

                               fixCommitRepo  \
0          https://github.com/apache/archiva   
1          https://github.com/apache/archiva   
2          https://github.com/apache/archiva   
3          https://github.com/apache/archiva   
4        https://github.com/x-stream/xstream   
..                                       ...   
645            https://github.com/oblac/jodd   
646            https://github.com/oblac/jodd   
647            https://github.com/oblac/jodd   
648  https://github.com/OWASP/json-sanitizer   
649  https://github.com/OWASP/json-sanitizer   

                                fixCommitHash  \
0    796716d44183bd315dd20184a66b39ae533eb747   
1    796716d44183bd315dd20184a66b39ae533eb747   
2    796716d44183bd315dd20184a66b39ae533eb747   
3    796716d44183bd315dd20184a66b39ae533eb747   
4    6344867dce6767af7d0fe34fb393271a6456672d   
..                                        ...   
645  9bffc3913aeb8472c11bb543243004b4b4376f16   
646  9bffc3913aeb8472c11bb543243004b4b4376f16   
647  9bffc3913aeb8472c11bb543243004b4b4376f16   
648  f3512bab078af3f5c533dacc2d410add8c0c1124   
649  f3512bab078af3f5c533dacc2d410add8c0c1124   

                            commitHashFromGit  \
0    796716d44183bd315dd20184a66b39ae533eb747   
1    796716d44183bd315dd20184a66b39ae533eb747   
2    796716d44183bd315dd20184a66b39ae533eb747   
3    796716d44183bd315dd20184a66b39ae533eb747   
4    6344867dce6767af7d0fe34fb393271a6456672d   
..                                        ...   
645  9bffc3913aeb8472c11bb543243004b4b4376f16   
646  9bffc3913aeb8472c11bb543243004b4b4376f16   
647  9bffc3913aeb8472c11bb543243004b4b4376f16   
648  f3512bab078af3f5c533dacc2d410add8c0c1124   
649  f3512bab078af3f5c533dacc2d410add8c0c1124   

                                modifiedFilePathBefore  \
0    archiva-modules/archiva-base/archiva-repositor...   
1    archiva-modules/archiva-base/archiva-repositor...   
2    archiva-modules/archiva-web/archiva-rest/archi...   
3    archiva-modules/archiva-web/archiva-rest/archi...   
4    xstream/src/java/com/thoughtworks/xstream/XStr...   
..                                                 ...   
645  jodd-json/src/main/java/jodd/json/JsonParserBa...   
646   jodd-json/src/main/java/jodd/json/MapToBean.java   
647  jodd-json/src/test/java/jodd/json/JSONDeserial...   
648   src/main/java/com/google/json/JsonSanitizer.java   
649  src/test/java/com/google/json/JsonSanitizerTes...   

                        

For 11 records, the file has been deleted.

In [17]:
df2 = df2[~df2['modifiedFilePathBefore'].isnull()]
df2 = df2[~df2['modifiedFileSrcBefore'].isnull()]
df2

cveId               ghsaId                       packageName  \
0     CVE-2019-0214  GHSA-jxgm-9f58-w4xp        org.apache.archiva:archiva   
1     CVE-2019-0214  GHSA-jxgm-9f58-w4xp        org.apache.archiva:archiva   
2     CVE-2019-0214  GHSA-jxgm-9f58-w4xp        org.apache.archiva:archiva   
3     CVE-2019-0214  GHSA-jxgm-9f58-w4xp        org.apache.archiva:archiva   
4     CVE-2013-7285  GHSA-f554-x222-wgf7  com.thoughtworks.xstream:xstream   
..              ...                  ...                               ...   
645  CVE-2018-21234  GHSA-jrg3-qq99-35g7                 org.jodd:jodd-all   
646  CVE-2018-21234  GHSA-jrg3-qq99-35g7                 org.jodd:jodd-all   
647  CVE-2018-21234  GHSA-jrg3-qq99-35g7                 org.jodd:jodd-all   
648  CVE-2020-13973  GHSA-g8jj-899q-8x3j     com.mikesamuel:json-sanitizer   
649  CVE-2020-13973  GHSA-g8jj-899q-8x3j     com.mikesamuel:json-sanitizer   

    vulnIntroduced latestVulnVersion vulnFixed  \
0            2.2.0             2.2.3     2.2.4   
1            2.2.0             2.2.3     2.2.4   
2            2.2.0             2.2.3     2.2.4   
3            2.2.0             2.2.3     2.2.4   
4                0             1.4.5     1.4.6   
..             ...               ...       ...   
645              0             5.0.3     5.0.4   
646              0             5.0.3     5.0.4   
647              0             5.0.3     5.0.4   
648              0             1.2.0     1.2.1   
649              0             1.2.0     1.2.1   

                               fixCommitRepo  \
0          https://github.com/apache/archiva   
1          https://github.com/apache/archiva   
2          https://github.com/apache/archiva   
3          https://github.com/apache/archiva   
4        https://github.com/x-stream/xstream   
..                                       ...   
645            https://github.com/oblac/jodd   
646            https://github.com/oblac/jodd   
647            https://github.com/oblac/jodd   
648  https://github.com/OWASP/json-sanitizer   
649  https://github.com/OWASP/json-sanitizer   

                                fixCommitHash  \
0    796716d44183bd315dd20184a66b39ae533eb747   
1    796716d44183bd315dd20184a66b39ae533eb747   
2    796716d44183bd315dd20184a66b39ae533eb747   
3    796716d44183bd315dd20184a66b39ae533eb747   
4    6344867dce6767af7d0fe34fb393271a6456672d   
..                                        ...   
645  9bffc3913aeb8472c11bb543243004b4b4376f16   
646  9bffc3913aeb8472c11bb543243004b4b4376f16   
647  9bffc3913aeb8472c11bb543243004b4b4376f16   
648  f3512bab078af3f5c533dacc2d410add8c0c1124   
649  f3512bab078af3f5c533dacc2d410add8c0c1124   

                            commitHashFromGit  \
0    796716d44183bd315dd20184a66b39ae533eb747   
1    796716d44183bd315dd20184a66b39ae533eb747   
2    796716d44183bd315dd20184a66b39ae533eb747   
3    796716d44183bd315dd20184a66b39ae533eb747   
4    6344867dce6767af7d0fe34fb393271a6456672d   
..                                        ...   
645  9bffc3913aeb8472c11bb543243004b4b4376f16   
646  9bffc3913aeb8472c11bb543243004b4b4376f16   
647  9bffc3913aeb8472c11bb543243004b4b4376f16   
648  f3512bab078af3f5c533dacc2d410add8c0c1124   
649  f3512bab078af3f5c533dacc2d410add8c0c1124   

                                modifiedFilePathBefore  \
0    archiva-modules/archiva-base/archiva-repositor...   
1    archiva-modules/archiva-base/archiva-repositor...   
2    archiva-modules/archiva-web/archiva-rest/archi...   
3    archiva-modules/archiva-web/archiva-rest/archi...   
4    xstream/src/java/com/thoughtworks/xstream/XStr...   
..                                                 ...   
645  jodd-json/src/main/java/jodd/json/JsonParserBa...   
646   jodd-json/src/main/java/jodd/json/MapToBean.java   
647  jodd-json/src/test/java/jodd/json/JSONDeserial...   
648   src/main/java/com/google/json/JsonSanitizer.java   
649  src/test/java/com/google/json/JsonSanitizerTes...   

                        

For 0 records, new file has been introduced.
639 records have been saved

In [18]:
df2.to_json(DATA_DIR/'interim'/'changes_from_git.json', orient='records', indent=2)

In [15]:
df2.columns.tolist()

['cveId',
 'ghsaId',
 'packageName',
 'vulnIntroduced',
 'latestVulnVersion',
 'vulnFixed',
 'fixCommitRepo',
 'fixCommitHash',
 'commitHashFromGit',
 'modifiedFilePathBefore',
 'modifiedFilePathAfter',
 'modifiedFileSrcBefore',
 'modifiedFileSrcAfter',
 'diffParsedJson',
 'nloc',
 'changedMethods']